File: src/com/douma/_13_day/_170/_170_TwoSum2.go
type TwoSum struct {
    nums map[int]int
}


func Constructor() TwoSum {
    nums := make(map[int]int)
    return TwoSum{nums:nums}
}


func (this *TwoSum) Add(number int)  {
    this.nums[number]++
}


func (this *TwoSum) Find(value int) bool {
    for key, _ := range this.nums {
        var target = value - key
        if target == key && this.nums[target] > 1 {
            return true
        }
        if target != key && this.nums[target] > 0 {
            return true
        }
    }

    return false
}


/**
 * Your TwoSum object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Add(number);
 * param_2 := obj.Find(value);
 */

File: src/com/douma/_13_day/_170/_170_TwoSum2.py
class TwoSum:

    def __init__(self):
        self.nums = dict()

    def add(self, number: int) -> None:
        if number in self.nums:
            self.nums[number] += 1
        else:
            self.nums[number] = 1


    def find(self, value: int) -> bool:
        for key in self.nums.keys():
            target = value - key
            if target == key and self.nums[target] > 1:
                return True
            if target != key and target in self.nums and self.nums[target] > 0:
                return True
        return False


# Your TwoSum object will be instantiated and called as such:
# obj = TwoSum()
# obj.add(number)
# param_2 = obj.find(value)